Refactor Linter / Packaging for Linter

robshakir opened the pull request:
@aashaikh 

PTAL. This is a major restructure of the repo - plus refactor of the linter, and addition of somewhat exhaustive unit tests for the linter. At this point, it has good coverage of all the style guide rules.

There are a couple of things that need to be tested here around the docs plugin, so I'd like to discuss that.

robshakir committed:
Refactor OpenConfig linter for flexibility and add testing.

  * (M) plugins/openconfig.py
  	- Refactor OC linter using an approach whereby a new mapping
  	  class is used to map statements to validation functions.
  	- Ensure compliance with pylint.
  * (M) plugins/util/yangpath.py
  	- Ensure compliance with pylint.
  * (A) requirements.txt
  	- Freeze the requirements for running OpenConfig linter.
  * (A) tests/oclinter:
  	- Add tests per-validation rule for the linter - to be used in CI.

robshakir committed:
Merge branch 'master' into oclinter-refactor

robshakir committed:
Major repository restructure and packaging changes.

	* (M) *
	  - Restructure the repo to support packaging, now can be built as
	    a PyPi package and distributed. Intention is to support CI.

aashaikh commented on openconfig_pyang/plugins/openconfig.py:
```diff
@@ -0,0 +1,893 @@
+                 "RFC 6087: 4.7"),
+      "submodule": (("contact", "organization", "description", "revision"),
+                    "RFC 6087: 4.7"),
+      # This rule is explicitly disabled because it is violated by
```
might it be better to have a module-level exception for this rather than turn off the check altogether for `iana-if-types`?  The description check in 4.12 is a useful reminder, for example.   Are we checking for these some other way?

robshakir replied:
ACK. I have removed these explicitly disabled rules, and added module level overrides for both `ietf-interfaces` and `iana-if-type`. I'll check in the OC CI that this continues to work as expected, also added unit tests for these.

aashaikh commented on tests/oclinter/enum-style/openconfig-testcase-succeed.yang:
```diff
@@ -0,0 +1,39 @@
+  grouping testcase-config {
+    leaf enumeration {
+      type enumeration {
+        enum WITH_FLOORLINES;
```
underscores ?

robshakir replied:
Done :-)

aashaikh commented on tests/oclinter/identity-case/openconfig-testcase-fail.yang:
```diff
@@ -0,0 +1,25 @@
+      "Placeholder";
+  }
+
+  identity lowercase;
```
for both enums and identities, do we need tests for some of other cases we've discovered (e.g., hex values with mix lower/upper case) ?

robshakir replied:
Done. As mentioned it doesn't look like 6020 has any restriction on the first character being numeric.

aashaikh commented on tests/oclinter/list-in-config-state/openconfig-testcase-fail.yang:
```diff
@@ -0,0 +1,53 @@
+  }
+
+  grouping testcase-config {
+    list list {
```
name this list test-list for clarity?

robshakir replied:
Done.

aashaikh commented on tests/oclinter/module-toplevel/openconfig-testcase-fail.yang:
```diff
@@ -0,0 +1,35 @@
+  import openconfig-extensions { prefix oc-ext; }
+
+  description
+    "Simple testcase to check compliance with quoting
```
a nit, but this description seems to be the same for many of the tests, even those not checking string quoting.  something to fix later perhaps.

aashaikh commented on tests/oclinter/paths-relative-leafref/openconfig-testcase-succeed.yang:
```diff
@@ -0,0 +1,50 @@
+module openconfig-testcase-succeed {
```
not sure i understand this test -- should it include a leafref type with a relative path?

robshakir replied:
Yes, this was an error. Test corrected.

aashaikh reviewed (commented):
thanks for this _huge_ amount of work!

Looks great, just a few minor comments / questions.  I probably need to ask you some questions f2f to understand the structure a bit better.

aashaikh commented:
also wanted to understand the reason we moved all the code into openconfig_plugins/ from plugins (to enable the pip packaging?)

may also want to double check we have a license at the top of each of the files (e.g., setup.py).

robshakir committed:
Correct issues within PR review.

  * (M) openconfig_pyang/__init__:
     - Update version number
  * (M) setup.py
     - Add copyright notice.
  * (M) tests/*
     - Update all YANG model descriptions to reflect the up-to-date values.

robshakir commented:
All files that have useful content have a copyright notice in them now (e.g., the version does not). 

PTAL at the edits :-)

aashaikh commented on tests/oclinter/enum-case/openconfig-testcase-succeed.yang:
```diff
@@ -5,8 +5,8 @@ module openconfig-testcase-succeed {
   description
-    "Simple testcase to check compliance with quoting
-    of strings";
+    "Suceed test case for case of enumeration in an
```
s/suceed/succeed

aashaikh reviewed (approved):
LGTM.

noticed one minor nit

robshakir committed:
Fix spelling error in enum-case test YANG.

  * (M) tests/oclinter/enum-case/openconfig-testcase-succeed.yang
    - Correct spelling error.

robshakir commented:
Thanks for the detailed review!
